### DISCUSSION TITLE
{today.date} not working in tcsh but working in zsh

### DISCUSSION THREAD

**Discussion by PDHell** [RhetTbull/osxphotos#1531] [Q&A]
Hello,

My MacOS's default shell is zsh. But since I am comfortable with tcsh, I changed my default shell to tcsh. 

My question is two fold. Could you please help. 

#1. 
When I use below script, it is working perfectly and I see `exportlog_2024-04-22_01H.27M.41S.csv` file from `--report`
```
#!/bin/zsh
osxphotos \
export /Users/karthikd/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL \
--library /Users/karthikd/Pictures/Photos\ Library.photoslibrary \
--filename "{created.date}.{original_name}{favorite?.Fav,}" \
--report ~/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL/exportlog_{today.date}_{today.hour}H.{today.min}M.{today.sec}S.csv \
--save-config /Users/karthikd/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL/osxphotos.toml \
--update \
--cleanup \
--export-by-date \
--touch-file \
--ramdb  \
--verbose \
--not-hidden \
--keyword-template "{label,KarNoLabel}" \
--keyword-template "{keyword,KarNoKeyword}" \
--keyword-template "{album?KarAlbumName:{folder_album}}" \
--keyword-template "{person}" \
--keyword-template "{favorite?Favorite,}" \
--keyword-template "{descr}" \
--person-keyword \
--xattr-template findercomment "Persons = {, +person,NoPerson} ; Place = {place.name,NoPlace}; Labels = {, +label,NoLabel}; Favorite={favorite?Yes,No}; Keywords = {, +keyword,NoKeyword} ; Album = {, +folder_album,NoAlbum} ; Description = {descr,NoDescr} " \
--keep "/*.csv"  \
--keep "/*.toml"  \
--keep "/*.log"  \
--sidecar XMP --from-date "2024-04-01"
```

But when execute below script, my --report file looks like this: `exportlog_today.date_today.hourH.today.minM.today.secS`
Looks like variables are not substituting correctly. 

Is it expected that variables replaced only in `zsh` but not in `tcsh` ?

```
#!/bin/tcsh
osxphotos \
export /Users/karthikd/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL \
--library /Users/karthikd/Pictures/Photos\ Library.photoslibrary \
--filename "{created.date}.{original_name}{favorite?.Fav,}" \
--report ~/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL/exportlog_{today.date}_{today.hour}H.{today.min}M.{today.sec}S.csv \
--save-config /Users/karthikd/Pictures/BackupOfPhotosLibraryUsingOSXPHOTOS_DEL/osxphotos.toml \
--update \
--cleanup \
--export-by-date \
--touch-file \
--ramdb  \
--verbose \
--not-hidden \
--keyword-template "{label,KarNoLabel}" \
--keyword-template "{keyword,KarNoKeyword}" \
--keyword-template "{album?KarAlbumName:{folder_album}}" \
--keyword-template "{person}" \
--keyword-template "{favorite?Favorite,}" \
--keyword-template "{descr}" \
--person-keyword \
--xattr-template findercomment "Persons = {, +person,NoPerson} ; Place = {place.name,NoPlace}; Labels = {, +label,NoLabel}; Favorite={favorite?Yes,No}; Keywords = {, +keyword,NoKeyword} ; Album = {, +folder_album,NoAlbum} ; Description = {descr,NoDescr} " \
--keep "/*.csv"  \
--keep "/*.toml"  \
--keep "/*.log"  \
--sidecar XMP --from-date "2024-04-01"
```

#2.
Also, is it okay to use `osxphotos` in `tcsh` in general ? Will I face any issues going forward? (I just started using `osxphotos`)

**Comment by RhetTbull** [comment]
The shell shouldn't matter. This has to do with quoting/escaping behavior. Try quoting the template for the report. But you'll probably have to list the whole path as ~ won't expand if quoted.

Yes you can use osxphotos with tcsh -- you just need to be aware of how your shell handles things like braces.

**Reply by PDHell** [reply]
Got it. Thank you so much!

Also, can I use variables in `--report` ?
For example, can I use below:
`--report       /Users/kar/Pictures/OSXPhotosBackup/{today.date}__{today.hour}h.{today.min}m.{today.sec}s__report.csv \
`

The reason I want to do this is: For every run I want to know the command I to run that particualt set. Basically everytime I run, I want to dump the command I use into some file for reference. Please let me know if there is a better way. 

As of now I have below:
```
--report       /Users/kar/Pictures/OSXPhotosBackup/{today.date}__{today.hour}h.{today.min}m.{today.sec}s__report.csv 
--save-config  /Users/kar/Pictures/OSXPhotosBackup/{today.date}__{today.hour}h.{today.min}m.{today.sec}s__report.toml 
```
And I am getting below two files. 
```
2024-04-22__20h.11m.03s__report.csv
{today.date}__{today.hour}h.{today.min}m.{today.sec}s__report.toml
```
So, my command for `--report` is working but `--save-config` is not working.

**Reply by RhetTbull** [reply]
Yes, the template variables can be used in `--report`. Unfortunately they cannot yet be used in `--save-config` though it might be good to add this. I'll open a new issue to allow `--save-config` to use template variables.

**Reply by PDHell** [reply]
Got it. In the meanwhile I will do like this:
`--save-config  /Users/karthikd/Pictures/OSXPhotosBackup/${datestamp}__config.toml \
`

One more question please:
As of now I am using this:
```
--keyword-template "{album?KarAlbumName:{folder_album},}" \
--keyword-template "{person?Person:{person},}" \
--keyword-template "{favorite?Favorite,}" \
--keyword-template "{label,KarNoLabel}" \
--keyword-template "{keyword,KarNoKeyword}" \
--keyword-template "{descr}" \
```
As you can see If there is an album, I am adding `<CustomText>:album_name`. Similarly adding custom texts for person. 

Can I extend this and add custom text for `label`, `keyword`, `Description`? Are there any fall outs of I do like below?
```
--keyword-template "{album?KarAlbumName:{folder_album},}" \
--keyword-template "{person?Person:{person},}" \
--keyword-template "{favorite?Favorite,}" \
--keyword-template "{label?KarLabel:{label},KarNoLabel}" \
--keyword-template "{keyword?KarKeyword:{keyword},KarNoKeyword}" \
--keyword-template "{descr?KarDescription:{descr}}" \
```


Also do I have to use `KarAlbumName:` or can I simply use `KarAlbumName=` ?
`=` is easy to query on XML file if needed.

**Reply by RhetTbull** [reply]
> Can I extend this and add custom text for label, keyword, Description? Are there any fall outs of I do like below?

Those should work fine.  

> Also do I have to use KarAlbumName: or can I simply use KarAlbumName= ?

This should work fine: `{album?KarAlbumName={folder_album},}`

I don't believe that `=` in the keyword will cause any issues.

You can use the `osxphotos template` command to test your templates.  Select a photo in Photos then run `osxphotos template` and type your tempaltes (for example, `{label?KarLabel:{label},KarNoLabel}` and you'll see it rendered for the currently selected photo. If you switch the selection and render the template again (hit up-arrow to recall the last template) it will render for the new photo. This is useful for debugging and testing template strings.
